File: src/test/java/com/jinloes/prpilot/services/ClaudeServiceTest.java
package com.jinloes.prpilot.services;

import static org.assertj.core.api.Assertions.assertThat;

import com.jinloes.prpilot.model.PRReviewRequest;
import com.jinloes.prpilot.model.PullRequest;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

class ClaudeServiceTest {

    private static PullRequest pr(String body) {
        return new PullRequest("My PR", "", "owner", "repo", 42, body, "author", "2024-01-01");
    }

    @Nested
    class BuildPrompt {

        @Test
        void alwaysContainsPersonaAndDiff() {
            String prompt =
                    ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff content", "", ""));
            assertThat(prompt).contains("senior security engineer");
            assertThat(prompt).contains("diff content");
        }

        @Test
        void diffWrappedInXmlTags() {
            String prompt =
                    ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff content", "", ""));
            assertThat(prompt).contains("<diff>\ndiff content\n</diff>");
        }

        @Test
        void blankProjectConventions_sectionAbsent() {
            String prompt = ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff", "", ""));
            // When blank, <project_conventions> must NOT appear as a block wrapper (with newline);
            // the instruction text references these tags inline (without a trailing newline).
            assertThat(prompt).doesNotContain("<project_conventions>\n");
        }

        @Test
        void nonBlankProjectConventions_wrappedInXmlTagsBeforeDiff() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr(""), "diff", "", "no magic numbers"));
            assertThat(prompt)
                    .contains("<project_conventions>\nno magic numbers\n</project_conventions>");
            // Use the content-carrying form to locate the injected block, not the inline mention
            assertThat(prompt.indexOf("<project_conventions>\nno magic numbers"))
                    .isLessThan(prompt.indexOf("<diff>\ndiff\n</diff>"));
        }

        @Test
        void blankKnownPatterns_sectionAbsent() {
            String prompt = ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff", "", ""));
            // When blank, <known_patterns> must NOT appear as a block wrapper (with newline)
            assertThat(prompt).doesNotContain("<known_patterns>\n");
        }

        @Test
        void nonBlankKnownPatterns_wrappedInXmlTags() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr(""), "diff", "use Optional", ""));
            assertThat(prompt).contains("<known_patterns>\n");
            assertThat(prompt).contains("</known_patterns>");
            assertThat(prompt).contains("use Optional");
        }

        @Test
        void conventionsSectionAppearsBeforeKnownPatterns() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr(""), "diff", "patterns", "conventions"));
            // Use the content-carrying form to locate the injected blocks
            assertThat(prompt.indexOf("<project_conventions>\nconventions"))
                    .isLessThan(prompt.indexOf("<known_patterns>\n"));
        }

        @Test
        void blankPrBody_descriptionSectionAbsent() {
            String prompt = ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff", "", ""));
            // When blank, <pr_description> must NOT appear as a block wrapper (with newline)
            assertThat(prompt).doesNotContain("<pr_description>\n");
        }

        @Test
        void nonBlankPrBody_wrappedInXmlTags() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr("fixes the bug"), "diff", "", ""));
            assertThat(prompt).contains("<pr_description>\nfixes the bug\n</pr_description>");
        }

        @Test
        void noPriorReview_sectionAbsent() {
            String prompt = ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff", "", ""));
            assertThat(prompt).doesNotContain("<prior_review>\n");
        }

        @Test
        void blankPriorReview_sectionAbsent() {
            String prompt =
                    ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff", "", "", "   "));
            assertThat(prompt).doesNotContain("<prior_review>\n");
        }

        @Test
        void nonBlankPriorReview_wrappedInXmlTags() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr(""), "diff", "", "", "Verdict: APPROVE"));
            assertThat(prompt).contains("<prior_review>\n");
            assertThat(prompt).contains("</prior_review>");
            assertThat(prompt).contains("Verdict: APPROVE");
        }

        @Test
        void priorReviewAppearsAfterKnownPatternsBeforePrDescription() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(
                                    pr("body"), "diff", "patterns", "conventions", "prior"));
            // Use content-carrying forms to avoid matching inline tag mentions in
            // REVIEW_INSTRUCTIONS
            int knownEnd = prompt.indexOf("</known_patterns>");
            int priorStart = prompt.indexOf("<prior_review>\n");
            int descStart = prompt.indexOf("<pr_description>\nbody");
            assertThat(knownEnd).isLessThan(priorStart);
            assertThat(priorStart).isLessThan(descStart);
        }

        @Test
        void priorReviewStripped() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr(""), "diff", "", "", "  spaced  "));
            assertThat(prompt).contains("<prior_review>\n");
            // The content should be stripped of surrounding whitespace
            assertThat(prompt).contains("spaced");
            assertThat(prompt).doesNotContain("  spaced  ");
        }

        @Test
        void noExistingReviews_sectionAbsent() {
            String prompt = ClaudeService.buildPrompt(new PRReviewRequest(pr(""), "diff", "", ""));
            assertThat(prompt).doesNotContain("<existing_reviews>\n");
        }

        @Test
        void blankExistingReviews_sectionAbsent() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr(""), "diff", "", "", null, "   "));
            assertThat(prompt).doesNotContain("<existing_reviews>\n");
        }

        @Test
        void nonBlankExistingReviews_wrappedInXmlTags() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(pr(""), "diff", "", "", null, "Review by @alice"));
            assertThat(prompt).contains("<existing_reviews>\n");
            assertThat(prompt).contains("</existing_reviews>");
            assertThat(prompt).contains("Review by @alice");
        }

        @Test
        void existingReviewsAppearsAfterKnownPatternsBeforePriorReview() {
            String prompt =
                    ClaudeService.buildPrompt(
                            new PRReviewRequest(
                                    pr("body"),
                                    "diff",
                                    "patterns",
                                    "conventions",
                                    "prior",
                                    "existing"));
            int knownEnd = prompt.indexOf("</known_patterns>");
            int existingStart = prompt.indexOf("<existing_reviews>\n");
            int priorStart = prompt.indexOf("<prior_review>\n");
            assertThat(knownEnd).isLessThan(existingStart);
            assertThat(existingStart).isLessThan(priorStart);
        }
    }

    @Nested
    class CancelCurrentRequest {

        @Test
        void noActiveProcess_doesNotThrow() {
            ClaudeService service = new ClaudeService();
            // Should be a no-op when idle (activeProcess is null)
            service.cancelCurrentRequest();
        }
    }

    @Nested
    class ToolUseStatus {

        @Test
        void simpleToolName_formatsWithEmptyArgs() {
            assertThat(ClaudeService.toolUseStatus("my_tool", Map.of())).isEqualTo("my_tool()");
        }

        @Test
        void mcpPrefixStripped_andDoubleUnderscoreReplacedWithSlash() {
            assertThat(ClaudeService.toolUseStatus("mcp__github__get_file", Map.of()))
                    .isEqualTo("github/get_file()");
        }

        @Test
        void noMcpPrefix_doubleUnderscoreStillReplaced() {
            assertThat(ClaudeService.toolUseStatus("ns__tool", Map.of())).isEqualTo("ns/tool()");
        }

        @Test
        void primitiveArgs_includedInOutput() {
            Map<String, Object> input = Map.of("owner", "alice", "repo", "myrepo");
            String result = ClaudeService.toolUseStatus("mcp__github__search", input);
            assertThat(result).contains("owner=alice");
            assertThat(result).contains("repo=myrepo");
        }

        @Test
        void multipleArgs_joinedWithCommaSpace() {
            Map<String, Object> input = Map.of("a", "1", "b", "2");
            String result = ClaudeService.toolUseStatus("tool", input);
            // Both entries present, separated by ", "
            assertThat(result).matches("tool\\(.*=.*,\\s.*=.*\\)");
        }

        @Test
        void nonPrimitiveArg_excluded() {
            Map<String, Object> input = new HashMap<>();
            input.put("nested", Map.of());
            input.put("list", List.of());
            input.put("scalar", "val");
            String result = ClaudeService.toolUseStatus("tool", input);
            assertThat(result).isEqualTo("tool(scalar=val)");
        }

        @Test
        void pathContainsClaudeDir_returnsNull() {
            assertThat(
                            ClaudeService.toolUseStatus(
                                    "tool", Map.of("path", "/home/user/.claude/tmp/abc")))
                    .isNull();
        }

        @Test
        void filePathContainsClaudeDir_returnsNull() {
            assertThat(
                            ClaudeService.toolUseStatus(
                                    "tool", Map.of("file_path", "/home/user/.claude/settings")))
                    .isNull();
        }

        @Test
        void filenameContainsClaudeDir_returnsNull() {
            assertThat(
                            ClaudeService.toolUseStatus(
                                    "tool", Map.of("filename", "C:\\Users\\user\\.claude\\tmp")))
                    .isNull();
        }

        @Test
        void pathOutsideClaudeDir_notSuppressed() {
            assertThat(
                            ClaudeService.toolUseStatus(
                                    "tool", Map.of("path", "/home/user/projects/src/Foo.java")))
                    .isNotNull();
        }

        @Test
        void nullPathValue_notSuppressed() {
            Map<String, Object> input = new HashMap<>();
            input.put("path", null);
            assertThat(ClaudeService.toolUseStatus("tool", input)).isNotNull();
        }
    }
}

File: core/src/jvmTest/kotlin/com/jinloes/prpilot/services/ClaudeServiceTest.kt
package com.jinloes.prpilot.services

import com.jinloes.prpilot.model.ChatMessage
import com.jinloes.prpilot.model.PRReviewRequest
import com.jinloes.prpilot.model.PullRequest
import com.jinloes.prpilot.services.stream.ContentBlock
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import io.kotest.matchers.string.shouldStartWith

// ── helpers ────────────────────────────────────────────────────────────────

private fun pr(body: String = "") =
    PullRequest("My PR", "", "owner", "repo", 42, body, "author", "2024-01-01")

private fun textBlock(text: String) = ContentBlock().also { it.type = "text"; it.text = text }
private fun thinkingBlock(thinking: String) = ContentBlock().also { it.type = "thinking"; it.thinking = thinking }
private fun toolUseBlock(name: String) = ContentBlock().also { it.type = "tool_use"; it.name = name; it.input = emptyMap() }

// ─────────────────────────────────────────────────────────────────────────

class ClaudeServiceTest : FunSpec({

    // ── buildPrompt ──────────────────────────────────────────────────────

    context("buildPrompt") {

        test("contains persona and fetch instruction") {
            val p = PullRequest("Fix the bug", "", "myorg", "myrepo", 99, "", "alice", "2024-01-01")
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(p, "", ""))
            prompt shouldContain "experienced engineer"
            prompt shouldContain "gh pr diff 99 --repo myorg/myrepo"
        }

        test("fetch diff wrapped in fetch_diff tag — no inline diff tag") {
            val p = PullRequest("Fix the bug", "", "myorg", "myrepo", 99, "", "alice", "2024-01-01")
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(p, "", ""))
            prompt shouldContain "<fetch_diff>\n"
            prompt shouldContain "gh pr diff 99 --repo myorg/myrepo"
            prompt shouldContain "</fetch_diff>"
            prompt shouldNotContain "<diff>"
        }

        test("pr_metadata wrapped in xml tags") {
            val p = PullRequest("Fix the bug", "", "myorg", "myrepo", 99, "", "alice", "2024-01-01")
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(p, "diff", ""))
            prompt shouldContain "<pr_metadata>\n"
            prompt shouldContain "</pr_metadata>"
            prompt shouldContain "number: 99"
            prompt shouldContain "repo: myorg/myrepo"
            prompt shouldContain "title: Fix the bug"
        }

        test("pr_metadata appears before fetch_diff") {
            val p = PullRequest("My PR", "", "org", "repo", 1, "", "alice", "2024-01-01")
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(p, "", ""))
            val metaIdx = prompt.indexOf("<pr_metadata>\nnumber:")
            val fetchIdx = prompt.indexOf("<fetch_diff>\nRun:")
            (metaIdx < fetchIdx) shouldBe true
        }

        test("blank knownPatterns — section absent") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", ""))
            prompt shouldNotContain "<known_patterns>\n"
        }

        test("non-blank knownPatterns — wrapped in xml tags") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", "use Optional"))
            prompt shouldContain "<known_patterns>\n"
            prompt shouldContain "</known_patterns>"
            prompt shouldContain "use Optional"
        }

        test("blank pr body — description section absent") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(""), "diff", ""))
            prompt shouldNotContain "<pr_description>\n"
        }

        test("non-blank pr body — wrapped in xml tags") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr("fixes the bug"), "diff", ""))
            prompt shouldContain "<pr_description>\nfixes the bug\n</pr_description>"
        }

        test("null priorReview — section absent") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", ""))
            prompt shouldNotContain "<prior_review>\n"
        }

        test("blank priorReview — section absent") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", "", "   "))
            prompt shouldNotContain "<prior_review>\n"
        }

        test("non-blank priorReview — wrapped in xml tags") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", "", "Verdict: APPROVE"))
            prompt shouldContain "<prior_review>\n"
            prompt shouldContain "</prior_review>"
            prompt shouldContain "Verdict: APPROVE"
        }

        test("priorReview appears after knownPatterns but before pr_description") {
            val prompt = ClaudeService.buildPrompt(
                PRReviewRequest(pr("body"), "diff", "patterns", "prior")
            )
            val knownEnd = prompt.indexOf("</known_patterns>")
            val priorStart = prompt.indexOf("<prior_review>\n")
            val descStart = prompt.indexOf("<pr_description>\nbody")
            (knownEnd < priorStart) shouldBe true
            (priorStart < descStart) shouldBe true
        }

        test("misattribution guard present") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "", ""))
            prompt shouldContain "misattributed comment is worse than no comment"
            prompt shouldContain "trace"
        }

        test("unverifiable issue guard present") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "", ""))
            prompt shouldContain "library internals"
            prompt shouldContain "When in doubt, leave it out"
        }

        test("issue type requires confirmation from diff") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "", ""))
            prompt shouldContain "Do NOT use \"issue\" for problems that require runtime"
        }

        test("prior_review tag listed in untrusted input section before actual content") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "", ""))
            val injectionGuardIdx = prompt.indexOf("untrusted input")
            val priorTagIdx = prompt.indexOf("<prior_review>", 0)
            (priorTagIdx < injectionGuardIdx) shouldBe true
        }

        test("line comments cap present") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "", ""))
            prompt shouldContain "at most 12 comments"
        }

        test("test coverage rule is scoped") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "", ""))
            prompt shouldContain "flag as \"issue\" only if"
            prompt shouldNotContain "any non-trivial new or changed branch or"
        }

        test("line numbering rules appear before schema") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "", ""))
            val lnIdx = prompt.indexOf("Line numbering:")
            val schemaIdx = prompt.indexOf("Schema (emit")
            (lnIdx < schemaIdx) shouldBe true
        }

        test("prior review content is stripped") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", "", "  spaced  "))
            prompt shouldContain "<prior_review>\n"
            prompt shouldContain "spaced"
            prompt shouldNotContain "  spaced  "
        }

        test("null existingReviews — section absent") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", ""))
            prompt shouldNotContain "<existing_reviews>\n"
        }

        test("blank existingReviews — section absent") {
            val prompt = ClaudeService.buildPrompt(PRReviewRequest(pr(), "diff", "", null, "   "))
            prompt shouldNotContain "<existing_reviews>\n"
        }

        test("non-blank existingReviews — wrapped in xml tags") {
            val prompt = ClaudeService.buildPrompt(
                PRReviewRequest(pr(), "diff", "", null, "Review by @alice")
            )
            prompt shouldContain "<existing_reviews>\n"
            prompt shouldContain "</existing_reviews>"
            prompt shouldContain "Review by @alice"
        }

        test("existingReviews appears after knownPatterns but before priorReview") {
            val prompt = ClaudeService.buildPrompt(
                PRReviewRequest(pr("body"), "diff", "patterns", "prior", "existing")
            )
            val knownEnd = prompt.indexOf("</known_patterns>")
            val existingStart = prompt.indexOf("<existing_reviews>\n")
            val priorStart = prompt.indexOf("<prior_review>\n")
            (knownEnd < existingStart) shouldBe true
            (existingStart < priorStart) shouldBe true
        }
    }

    // ── cancelCurrentRequest ─────────────────────────────────────────────

    context("cancelCurrentRequest") {

        test("no active process — does not throw") {
            val service = ClaudeService()
            service.cancelCurrentRequest() // should not throw
        }
    }

    // ── handleContentBlock ───────────────────────────────────────────────

    context("handleContentBlock") {
        val service = ClaudeService()

        test("text block with onChunk — calls onChunk, not onStatus") {
            val statuses = mutableListOf<String>()
            val chunks = mutableListOf<Array<String>>()
            service.handleContentBlock(
                textBlock("hello world"),
                { statuses.add(it) },
                { k, v -> chunks.add(arrayOf(k, v)) },
            )
            statuses.shouldBeEmpty()
            chunks shouldHaveSize 1
            chunks[0][0] shouldBe "text"
            chunks[0][1] shouldBe "hello world"
        }

        test("text block with textBuffer — appends text") {
            val textBuffer = StringBuilder()
            service.handleContentBlock(textBlock("json content"), {}, null, textBuffer)
            textBuffer.toString() shouldBe "json content"
        }

        test("text block with blank text — does not append to textBuffer") {
            val textBuffer = StringBuilder()
            service.handleContentBlock(textBlock("  "), {}, null, textBuffer)
            textBuffer.toString() shouldBe ""
        }

        test("text block without onChunk — calls onStatus with 'Generating review…'") {
            val statuses = mutableListOf<String>()
            service.handleContentBlock(textBlock("hello"), { statuses.add(it) }, null)
            statuses shouldBe listOf("Generating review…")
        }

        test("text block with blank text and onChunk — falls back to onStatus") {
            val statuses = mutableListOf<String>()
            val chunks = mutableListOf<Array<String>>()
            service.handleContentBlock(
                textBlock("   "),
                { statuses.add(it) },
                { k, v -> chunks.add(arrayOf(k, v)) },
            )
            chunks.shouldBeEmpty()
            statuses shouldBe listOf("Generating review…")
        }

        test("thinking block with onChunk — calls onChunk with 'thinking'") {
            val chunks = mutableListOf<Array<String>>()
            service.handleContentBlock(
                thinkingBlock("deep thought"),
                {},
                { k, v -> chunks.add(arrayOf(k, v)) },
            )
            chunks shouldHaveSize 1
            chunks[0][0] shouldBe "thinking"
            chunks[0][1] shouldBe "deep thought"
        }

        test("thinking block without onChunk — nothing called") {
            val statuses = mutableListOf<String>()
            service.handleContentBlock(thinkingBlock("deep thought"), { statuses.add(it) }, null)
            statuses.shouldBeEmpty()
        }

        test("tool_use block — always calls onStatus") {
            val statuses = mutableListOf<String>()
            service.handleContentBlock(toolUseBlock("my_tool"), { statuses.add(it) }, { _, _ -> })
            statuses shouldBe listOf("my_tool()")
        }

        test("unknown block type — no exception and no callback") {
            val statuses = mutableListOf<String>()
            val block = ContentBlock().also { it.type = "unknown_future_type" }
            service.handleContentBlock(block, { statuses.add(it) }, null)
            statuses.shouldBeEmpty()
        }

        test("null type block — treated as empty string, no exception") {
            val statuses = mutableListOf<String>()
            val block = ContentBlock() // type is null by default
            service.handleContentBlock(block, { statuses.add(it) }, null)
            // no crash, defaultString("") falls through the when
            statuses.shouldBeEmpty()
        }
    }

    // ── toolUseStatus ────────────────────────────────────────────────────

    context("toolUseStatus") {

        test("simple tool name formats with empty args") {
            ClaudeService.toolUseStatus("my_tool", emptyMap()) shouldBe "my_tool()"
        }

        test("mcp__ prefix stripped and __ replaced with /") {
            ClaudeService.toolUseStatus("mcp__github__get_file", emptyMap()) shouldBe "github/get_file()"
        }

        test("no mcp prefix — double underscore still replaced") {
            ClaudeService.toolUseStatus("ns__tool", emptyMap()) shouldBe "ns/tool()"
        }

        test("primitive string args included in output") {
            val input = mapOf("owner" to "alice", "repo" to "myrepo")
            val result = ClaudeService.toolUseStatus("mcp__github__search", input)
            result shouldContain "owner=alice"
            result shouldContain "repo=myrepo"
        }

        test("multiple args joined with comma space") {
            val input = mapOf("a" to "1", "b" to "2")
            val result = ClaudeService.toolUseStatus("tool", input)
            // should match tool(<key>=<val>, <key>=<val>)
            result?.matches(Regex("tool\\(.*=.*,\\s.*=.*\\)")) shouldBe true
        }

        test("non-primitive args excluded") {
            val input: Map<String, Any> = mapOf(
                "nested" to emptyMap<String, Any>(),
                "list" to emptyList<Any>(),
                "scalar" to "val",
            )
            ClaudeService.toolUseStatus("tool", input) shouldBe "tool(scalar=val)"
        }

        test("path containing .claude/ returns null") {
            ClaudeService.toolUseStatus("tool", mapOf("path" to "/home/user/.claude/tmp/abc"))
                .shouldBeNull()
        }

        test("file_path containing .claude/ returns null") {
            ClaudeService.toolUseStatus("tool", mapOf("file_path" to "/home/user/.claude/settings"))
                .shouldBeNull()
        }

        test("filename containing .claude\\ (Windows) returns null") {
            ClaudeService.toolUseStatus("tool", mapOf("filename" to "C:\\Users\\user\\.claude\\tmp"))
                .shouldBeNull()
        }

        test("path outside .claude/ — not suppressed") {
            ClaudeService.toolUseStatus("tool", mapOf("path" to "/home/user/projects/src/Foo.java"))
                .shouldNotBeNull()
        }

        test("path value that is null — not suppressed") {
            val input: MutableMap<String, Any?> = mutableMapOf("path" to null)
            @Suppress("UNCHECKED_CAST")
            ClaudeService.toolUseStatus("tool", input as Map<String, Any>).shouldNotBeNull()
        }

        test("number arg included as scalar") {
            ClaudeService.toolUseStatus("tool", mapOf("count" to 42)) shouldBe "tool(count=42)"
        }

        test("boolean arg included as scalar") {
            ClaudeService.toolUseStatus("tool", mapOf("flag" to true)) shouldBe "tool(flag=true)"
        }
    }

    // ── buildChatPrompt ──────────────────────────────────────────────────

    context("buildChatPrompt") {

        test("user turn wrapped in user tag") {
            val history = listOf(ChatMessage(ChatMessage.Role.USER, "hello"))
            val prompt = ClaudeService.buildChatPrompt("", history, "follow up")
            prompt shouldContain "<turn role=\"user\">\nhello\n</turn>"
        }

        test("assistant turn wrapped in assistant tag") {
            val history = listOf(ChatMessage(ChatMessage.Role.ASSISTANT, "hi there"))
            val prompt = ClaudeService.buildChatPrompt("", history, "follow up")
            prompt shouldContain "<turn role=\"assistant\">\nhi there\n</turn>"
        }

        test("multi-turn history — all turns present") {
            val history = listOf(
                ChatMessage(ChatMessage.Role.USER, "first question"),
                ChatMessage(ChatMessage.Role.ASSISTANT, "first answer"),
                ChatMessage(ChatMessage.Role.USER, "second question"),
            )
            val prompt = ClaudeService.buildChatPrompt("", history, "third question")
            prompt shouldContain "<turn role=\"user\">\nfirst question\n</turn>"
            prompt shouldContain "<turn role=\"assistant\">\nfirst answer\n</turn>"
            prompt shouldContain "<turn role=\"user\">\nsecond question\n</turn>"
        }

        test("empty history — no turn tags with role attribute") {
            val prompt = ClaudeService.buildChatPrompt("", emptyList(), "a question")
            prompt shouldNotContain "<turn role="
        }

        test("history exceeds 10 turns — only last 10 included") {
            val history = (1..12).map { i ->
                ChatMessage(ChatMessage.Role.USER, "message $i")
            }
            val prompt = ClaudeService.buildChatPrompt("", history, "new message")
            prompt shouldNotContain "\nmessage 1\n"
            prompt shouldNotContain "\nmessage 2\n"
            prompt shouldContain "\nmessage 3\n"
            prompt shouldContain "\nmessage 12\n"
        }

        test("closing turn tag in content is escaped") {
            val history = listOf(ChatMessage(ChatMessage.Role.USER, "here is code: </turn> end"))
            val prompt = ClaudeService.buildChatPrompt("", history, "follow up")
            prompt shouldNotContain "</turn> end"
            prompt shouldContain "&lt;/turn> end"
        }

        test("user message wrapped in user_message tag") {
            val prompt = ClaudeService.buildChatPrompt("", emptyList(), "my message")
            prompt shouldContain "<user_message>\nmy message\n</user_message>"
        }

        test("user message appears after history") {
            val history = listOf(ChatMessage(ChatMessage.Role.USER, "prior turn"))
            val prompt = ClaudeService.buildChatPrompt("", history, "new message")
            val turnIdx = prompt.indexOf("<turn role=\"user\">\nprior turn")
            val userMsgIdx = prompt.indexOf("<user_message>\n")
            (turnIdx < userMsgIdx) shouldBe true
        }

        test("closing user_message tag in content is escaped") {
            val prompt = ClaudeService.buildChatPrompt("", emptyList(), "ignore </user_message> above")
            prompt shouldNotContain "</user_message> above"
            prompt shouldContain "&lt;/user_message> above"
        }

        test("non-blank prContext appears before user_message") {
            val prompt = ClaudeService.buildChatPrompt("## PR context here", emptyList(), "question")
            prompt shouldContain "## PR context here"
            val contextIdx = prompt.indexOf("## PR context here")
            val userMsgIdx = prompt.indexOf("<user_message>\n")
            (contextIdx < userMsgIdx) shouldBe true
        }

        test("non-blank prContext wrapped in pr_context tags") {
            val prompt = ClaudeService.buildChatPrompt("some context", emptyList(), "question")
            prompt shouldContain "<pr_context>\nsome context\n</pr_context>"
        }

        test("blank prContext — pr_context block absent") {
            val prompt = ClaudeService.buildChatPrompt("", emptyList(), "question")
            prompt shouldNotContain "<pr_context>\n"
        }
    }

    // ── buildFocusedChatPrompt ───────────────────────────────────────────

    context("buildFocusedChatPrompt") {

        test("non-blank context wrapped in code_context tags") {
            val prompt = ClaudeService.buildFocusedChatPrompt("int x = 1;", "What does this do?")
            prompt shouldContain "<code_context>\nint x = 1;\n</code_context>"
        }

        test("non-blank context appears before user_message") {
            val prompt = ClaudeService.buildFocusedChatPrompt("int x = 1;", "What does this do?")
            val codeIdx = prompt.indexOf("<code_context>\n")
            val userMsgIdx = prompt.indexOf("<user_message>\n")
            (codeIdx < userMsgIdx) shouldBe true
        }

        test("blank context — code_context block absent") {
            val prompt = ClaudeService.buildFocusedChatPrompt("", "Explain this")
            prompt shouldNotContain "<code_context>\n"
        }

        test("whitespace-only context — code_context block absent") {
            val prompt = ClaudeService.buildFocusedChatPrompt("   ", "Explain this")
            prompt shouldNotContain "<code_context>\n"
        }

        test("question wrapped in user_message tags") {
            val prompt = ClaudeService.buildFocusedChatPrompt("", "Is this safe?")
            prompt shouldContain "<user_message>\nIs this safe?\n</user_message>"
        }

        test("question always present even without context") {
            val prompt = ClaudeService.buildFocusedChatPrompt("", "My question")
            prompt shouldContain "My question"
        }

        test("chat persona appears at start") {
            val prompt = ClaudeService.buildFocusedChatPrompt("code", "question")
            prompt shouldStartWith "You are a senior engineer familiar"
        }

        test("persona appears before code_context") {
            val prompt = ClaudeService.buildFocusedChatPrompt("code", "question")
            val seniorIdx = prompt.indexOf("senior engineer")
            val codeCtxIdx = prompt.indexOf("<code_context>")
            (seniorIdx < codeCtxIdx) shouldBe true
        }
    }
})
